packages/use-dataloader/src/DataLoaderProvider/index.js

```diff
@@ -0,0 +1,148 @@
+import PropTypes from 'prop-types'
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useRef,
+} from 'react'
+
+export const DataLoaderContext = createContext()
+
+const DataLoaderProvider = ({ children }) => {
+  const cachedData = useRef({})
+  const reloads = useRef({})
+
+  const setCachedData = useCallback(compute => {
+    if (typeof compute === 'function') {
+      cachedData.current = compute(cachedData.current)
+    } else {
+      cachedData.current = compute
+    }
+  }, [])
+
+  const setReloads = useCallback(compute => {
+    if (typeof compute === 'function') {
+      reloads.current = compute(reloads.current)
+    } else {
+      reloads.current = compute
+    }
+  }, [])
+
+  const addCachedData = useCallback(
+    (key, newData) => {
+      if (key && typeof key === 'string' && newData) {
+        setCachedData(actualCachedData => ({
+          ...actualCachedData,
+          [key]: newData,
+        }))
+      }
+    },
+    [setCachedData],
+  )
+
+  const addReload = useCallback(
+    (key, method) => {
+      if (key && typeof key === 'string' && method) {
+        setReloads(actualReloads => ({
+          ...actualReloads,
+          [key]: method,
+        }))
+      }
+    },
+    [setReloads],
+  )
+
+  const clearReload = useCallback(
+    key => {
+      if (key && typeof key === 'string') {
+        setReloads(actualReloads => {
+          const tmp = actualReloads
+          delete tmp[key]
+
+          return tmp
+        })
+      } else {
+        setReloads({})
+      }
+    },
+    [setReloads],
+  )
+
+  const clearCachedData = useCallback(
+    key => {
+      if (key && typeof key === 'string') {
+        setCachedData(actualCachedData => {
+          const tmp = actualCachedData
+          delete tmp[key]
+
+          return tmp
+        })
+      } else {
+        setCachedData({})
+      }
+    },
+    [setCachedData],
+  )
+
+  const reload = useCallback(async key => {
+    if (key && typeof key === 'string') {
+      await (reloads.current[key] && reloads.current[key]())
+    } else {
+      await Promise.all(
+        Object.values(reloads.current).map(reloadFn => reloadFn()),
+      )
+    }
+  }, [])
+
+  const getCachedData = useCallback(key => {
+    if (key) {
+      return cachedData.current[key] || undefined
+    }
+
+    return cachedData.current
+  }, [])
+
+  const getReloads = useCallback(key => {
+    if (key) {
+      return reloads.current[key] || undefined
+    }
+
+    return reloads.current
+  }, [])
+
+  const value = useMemo(
+    () => ({
+      addCachedData,
```

Comment:
If I understand correctly most of this context should only be used internally ? 

I also don't really understand the value of `reload` vs `getReloads`, these functions basically have the same result no ?

Reply:
This context handle data cache and all function associated with their key. Like this you can trigger data reload from another component.  `reload` directly trigger a reload where as `getReloads` is just to retrieve a reload function so by exemple: 
```js
const dataloaderContext = useDataLoaderContex()
const reloadOtherComponent = () => {
  dataloaderContext.reload("test-key") // Reload only test-key function
  dataloaderContext.reload() // Reload all functions
}
// Not as easy as
const reloadOtherComponent = () => {
  dataloaderContext.getReloads("test-key")?.() // Reload test-key function
  dataloaderContext.getReloads().forEach(fn => fn()) // Reload all functions
}
```

But you can have a specific case where you need to get only the reload function, To ensure it exist or anything else